src/mscorlib/src/System/Double.cs

```diff
@@ -44,56 +44,87 @@ public struct Double : IComparable, IFormattable, IConvertible
 
         internal static double NegativeZero = BitConverter.Int64BitsToDouble(unchecked((long)0x8000000000000000));
 
+        /// <summary>Determines whether the specified value is finite (zero, subnormal, or normal).</summary>
         [Pure]
         [System.Runtime.Versioning.NonVersionable]
+        [MethodImpl(MethodImplOptions.AggressiveInlining)] // This compiles down to roughly 2 instructions
+        public unsafe static bool IsFinite(double d)
+        {
+            var bits = BitConverter.DoubleToInt64Bits(d);
+            return (bits & 0x7FFFFFFFFFFFFFFF) < 0x7FF0000000000000;
+        }
+
+        /// <summary>Determines whether the specified value is infinite.</summary>
+        [Pure]
+        [System.Runtime.Versioning.NonVersionable]
+        [MethodImpl(MethodImplOptions.AggressiveInlining)] // This compiles down to roughly 2 instructions
         public unsafe static bool IsInfinity(double d)
         {
-            return (*(long*)(&d) & 0x7FFFFFFFFFFFFFFF) == 0x7FF0000000000000;
+            var bits = BitConverter.DoubleToInt64Bits(d);
+            return (bits & 0x7FFFFFFFFFFFFFFF) == 0x7FF0000000000000;
         }
 
+        /// <summary>Determines whether the specified value is NaN.</summary>
         [Pure]
         [System.Runtime.Versioning.NonVersionable]
-        public static bool IsPositiveInfinity(double d)
+        [MethodImpl(MethodImplOptions.AggressiveInlining)] // This compiles down to roughly 2 instructions
+        public unsafe static bool IsNaN(double d)
         {
-            //Jit will generate inlineable code with this
-            if (d == double.PositiveInfinity)
-            {
-                return true;
-            }
-            else
-            {
-                return false;
-            }
+            var bits = BitConverter.DoubleToInt64Bits(d);
```

Comment:
> Codegen on x86 could be improved by only comparing the upper 32-bits return (upperbits & 0x7FFFFFFF) > 0x7FF00000.

It appears that you're using an older build or something, the current build produces:
```asm
G_M55887_IG02:
       C4E17B104508 vmovsd   xmm0, qword ptr [ebp+08H]
       C4E17B110424 vmovsd   qword ptr [esp], xmm0
       8B0424       mov      eax, dword ptr [esp]
       8B542404     mov      edx, dword ptr [esp+04H]
       83E0FF       and      eax, -1
       81E2FFFFFF7F and      edx, 0x7FFFFFFF
       83F801       cmp      eax, 1
       81DA0000F07F sbb      edx, 0x7FF00000
       0F9DC0       setge    al
       0FB6C0       movzx    eax, al
```